### ISSUE TITLE
Quality Drops and Identity Drifts for Manual Denoising Loop

### ISSUE THREAD

**Issue by AvirupJU** [foivospar/Arc2Face#33]
Hi, 
When the ID embeddings are passed to the SD pipeline the images are of good quality but when I use them in a manual denoising loop, the quality drops and there is a noticeable identity drift. Can you help me debug the issue?
 Here's is the code:
 ```
from diffusers import (
    StableDiffusionPipeline,
    UNet2DConditionModel,
    AutoencoderKL,
    DPMSolverMultistepScheduler,
)

from arc2face import CLIPTextModelWrapper, project_face_embs

import torch
from insightface.app import FaceAnalysis
from PIL import Image
import numpy as np
import os
import itertools
from tqdm import tqdm
base_model = 'stable-diffusion-v1-5/stable-diffusion-v1-5'

encoder = CLIPTextModelWrapper.from_pretrained(
    'models', subfolder="encoder", torch_dtype=torch.float16
).to('cuda')

unet = UNet2DConditionModel.from_pretrained(
    'models', subfolder="arc2face", torch_dtype=torch.float16
).to('cuda')

vae = AutoencoderKL.from_pretrained(base_model, subfolder="vae", torch_dtype=torch.float16).to("cuda")

pipeline = StableDiffusionPipeline.from_pretrained(
        base_model,
        text_encoder=encoder,
        unet=unet,
        torch_dtype=torch.float16,
        safety_checker=None
    ).to('cuda')

scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)

app = FaceAnalysis(name='antelopev2', root='./', providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
app.prepare(ctx_id=0, det_size=(256, 256))

def get_embedding(img):
    faces = app.get(img)
    faces = sorted(faces, key=lambda x:(x['bbox'][2]-x['bbox'][0])*(x['bbox'][3]-x['bbox'][1]))[-1]  # select largest face (if more than one detected)
    id_emb = torch.tensor(faces['embedding'], dtype=torch.float16)[None].cuda()
    id_emb = id_emb/torch.norm(id_emb, dim=1, keepdim=True)

    return id_emb.unsqueeze(0)

subject = 'obama'
guidance_scale = 3.0
num_images = 4
height, width = 512, 512  # Adjust to desired output size
img = np.array(Image.open(f'experiments/clusters/{subject}/{subject}.png'))[:,:,::-1]
num_inference_steps = 25

latents = torch.randn((num_images, unet.config.in_channels, height // 8, width // 8), device="cuda", dtype=torch.float16)
scheduler.set_timesteps(num_inference_steps=num_inference_steps)
id_emb = get_embedding(img)
id_emb = project_face_embs(pipeline, id_emb)
prompt_embeds = id_emb
prompt_embeds = prompt_embeds.repeat(num_images * 2, 1, 1) #batch duplicated latent's dimensions


for t in tqdm(scheduler.timesteps):
    
    latent_model_input = torch.cat([latents] * 2)  # Duplicate for guidance
    latent_model_input = scheduler.scale_model_input(latent_model_input, t)
    # Predict the noise
    with torch.no_grad():
        noise_pred = unet(latent_model_input, t, encoder_hidden_states=prompt_embeds).sample

    # Classifier-free guidance
    noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
    noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)

    # Update latents
    latents = scheduler.step(noise_pred, t, latents).prev_sample

# Decode latents to images using the VAE
latents = latents / 0.18215
with torch.no_grad():
    images = vae.decode(latents).sample 

# Post-process and save images
images = (images / 2 + 0.5).clamp(0, 1)  # Normalize images to [0, 1]
images = images.cpu().permute(0, 2, 3, 1).numpy()  # Convert to NumPy
# Save images
for idx, img in enumerate(images):
    img = Image.fromarray((img * 255).astype(np.uint8))
    img.save(f'experiments/clusters/{subject}/fake/{idx}_m.png')

```
Here are some images for reference. [Order: Seed, From SDPipeline, Manual]
<p align="center">
  <img src="https://github.com/user-attachments/assets/6832e62e-4d93-4985-9c23-32f36b56af75" alt="Image 1" width="30%">
  <img src="https://github.com/user-attachments/assets/af9a4a94-d1af-44cf-86a8-c435a475eecf" alt="Image 2" width="30%">
  <img src="https://github.com/user-attachments/assets/f646b5c5-29d0-46e4-a4bb-8be044493546" alt="Image 3" width="30%">
</p>

**Comment by AvirupJU** [comment]
The problem was with the prompt embedding. The prompt embedding must be passed into the ```pipeline.encode_prompt()``` method. The resulting positive and negative embedding should be concatenated and passed as an argument to the UNet(). Here's is the full code block:
```
prompt_embeds, negative_prompt_embeds = pipeline.encode_prompt(
            prompt='',
            device='cuda',
            num_images_per_prompt=num_images,
            do_classifier_free_guidance=True,
            negative_prompt='',
            prompt_embeds=id_emb,
            negative_prompt_embeds=None,
            lora_scale=None,
            clip_skip=None,
        )

prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds]).
```
Results after making the changes:
<p align="center">
  <img src="https://github.com/user-attachments/assets/2dcd7490-379b-4fad-9450-9579fd71038c" alt="Image 1" width="30%">
  <img src="https://github.com/user-attachments/assets/bbbdf0c0-3a35-4f59-a84f-507a0124bb59" alt="Image 2" width="30%">
</p>
